tests/puppeteer/pages/BO/addProduct.js

```diff
@@ -0,0 +1,191 @@
+    this.productCombinationBulkQuantityInput = '#product_combination_bulk_quantity';
+    this.productCombinationSelectAllCheckbox = 'input#toggle-all-combinations';
+    this.applyOnCombinationsButton = '#apply-on-combinations';
+    this.productCombinationTableRow = '#accordion_combinations tr:nth-of-type(%ID)';
+    this.deleteCombinationsButton = '#delete-combinations';
+    this.productCombinationsBulkForm = '#combinations-bulk-form';
+    this.productCombinationsBulkFormTitle= `${this.productCombinationsBulkForm} p[aria-controls='bulk-combinations-container']`;
+
+    // Growls : override value from BObasePage
+    this.growlMessageBloc = '#growls-default .growl-message';
+  }
+
+  /*
+  Methods
+   */
+  /**
+   * Create or edit product in BO
+   * @param productData
+   * @param switchProductOnline
+   * @return {Promise<textContent>}
+   */
+  async createEditProduct(productData, switchProductOnline = true) {
+    // Set Name, type of product, Reference, price ttc and quantity, and with combinations
+    await this.page.click(this.productNameInput, {clickCount: 3});
+    await this.page.type(this.productNameInput, productData.name);
+    await this.selectByVisibleText(this.productTypeSelect, productData.type);
+    await this.page.click(this.productReferenceInput, {clickCount: 3});
+    await this.page.type(this.productReferenceInput, productData.reference);
+    await this.page.click(this.productPriceTtcInput, {clickCount: 3});
+    await this.page.type(this.productPriceTtcInput, productData.price);
+    // Set description value
+    await this.page.click(this.productDescriotionTab);
+    await this.setValueOnTinymceInput(this.productDescriptionIframe, productData.description);
+    // Add combinations if exists
+    if (productData.withCombination) {
+      await this.page.click(this.productWithCombinationsInput);
+      await this.setCombinationsInProduct(productData);
+    }
+    else {
+      await this.page.click(this.productQuantityInput, {clickCount: 3});
+      await this.page.type(this.productQuantityInput, productData.quantity);
+    }
+    // Switch product online before save
+    if (switchProductOnline) {
+      await Promise.all([
+        this.page.waitForSelector(this.growlMessageBloc, {visible: true}),
+        this.page.click(this.productOnlineSwitch),
+      ]);
+    }
+    // Save created product
+    await Promise.all([
+      this.page.waitForSelector(this.growlMessageBloc, {visible: true}),
+      this.page.click(this.saveProductButton),
+    ]);
+    return this.getTextContent(this.growlMessageBloc);
+  }
+
+  /**
+   * Set Combinations for product
+   * @param productData
+   * @return {Promise<void>}
+   */
+  async setCombinationsInProduct(productData) {
+    // GOTO Combination tab : id = 3
+    await this.goToFormStep(3);
+    // Delete All combinations if exists
+    await this.deleteAllCombinations();
+    // set Combinations
+    const keys = Object.keys(productData.combinations);
+    /*eslint-disable*/
```

Comment:
👍

Reply:
yeah eslint don't like awaits in loop, and we need actions to be done sequentially (no way to use `Array.ForEach()`)